Rewrite all Skeleton logic

tafia opened the pull request:
I've reached a nice step in rewriting the spine library. While it is still not ready for integration, I wanted to share this milestone:
1. `skeleton` module to handle all logic from the skeleton
2. `Skeleton` struct: precomputed version of the json data (working with indices etc...)
3. `timeline` module: contains all the logic about interpolation between timelines. It also computes all bezier points in advance when necessary (for now, 10 points only between 2 frames, as done in other runtimes)
4. `animation` module: provides an iterator created per skin/animation and iterates over an `SRT` (renaming of `BoneData`) and attachment name (color is coming soon)
   - All the transformations are done on `SRT` instead of using Matrices (from one of my previous PR, I found that Matrix are expensive and the other runtimes don't use them)
   - I find iterators more rusty. If you want to loop animations, you can simply use the [`Cycle`](http://doc.rust-lang.org/std/iter/trait.Iterator.html#method.cycle) iterator for instance
5. So far I did not need crates like cgmath or color, even if I had to use rustc-serialize for hexadecimal color parsing ...

The current milestone is that **it all compiles**. 

I intentionally kept old logic as they just share the same json deserialization and it is simpler to compare

I haven't cross checked figures yet, first benches show a x5 improvement.
That's it. I suppose I'll spend the next days trying to build a tool to _see_ the animation and figure how wrong I am (I've read your glium tutorial and it could be enough)

tomaka commented on src/skeleton/animation.rs:
```diff
@@ -0,0 +1,113 @@
+        }
+
+        // increase time
+        self.time += 0.01f32;
```
Does this code assume that we're animating 10ms per 10ms?

tafia replied:
Yes (100Hz)
I plan to have a `delta` parameter on `AnimationIter` but I don't really know the practice in gaming (should we go for 60 Hz ?).
I hard-coded it and will add this aspect in my todo list as well

tomaka commented on src/skeleton/error.rs:
```diff
@@ -0,0 +1,37 @@
+    AnimationNotFound(String),
+}
+
+impl From<SkeletonError> for String {
```
`SkeletonError` should implement `Debug` (with a #[derive]), `Error` and `Display` (by calling `Error::description`).
But not `Into<String>`.

Example: https://github.com/tomaka/glium/blob/6713ac5f9843ceb64e2dd0fa6c5742b10203ae25/src/buffer/mod.rs#L174-L197

tafia replied:
Fine, I'll update

tomaka commented on src/skeleton/mod.rs:
```diff
@@ -0,0 +1,281 @@
+// Reexport skeleton modules
+pub use self::error::*;
+use self::timelines::*;
+use self::animation::*;
```
Ideally glob imports shouldn't be used.

tafia replied:
Yes I agree. At the beginning I didn't really know what exactly was supposed to be publicly exposed. I start having a better idea now most things are in place so I'll change this too

tomaka commented on src/skeleton/mod.rs:
```diff
@@ -0,0 +1,281 @@
+impl Skeleton {
+
+    /// Consumes reader (with json data) and returns a skeleton wrapping
+    pub fn from_reader<R: Read>(mut reader: R) -> Result<Skeleton, String> {
```
I suggest `Box<Error>` instead of `String` if you don't want to bother or don't want to expose the underlying error.

tafia replied:
Looks better indeed

tafia commented:
I have a huge [todo list](https://github.com/tafia/spine-rs/issues/1) but the global shape should be there. Thank you very much for the review!

tomaka commented:
There's no point in providing `from_file`, especially if it adds a new error type.

tafia commented:
Hmm
I thought it may help. But I agree it could be done outside.
On 18 Oct 2015 22:07, "tomaka" notifications@github.com wrote:

> There's no point in providing from_file, especially if it adds a new
> error type.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/tomaka/spine-rs/pull/6#issuecomment-149019395.

tomaka commented:
I just re-read your changes.
Everything looks pretty good! I have two remarks though:
- In my opinion it would be better if the animation iterator returned a struct instead of a tuple.
- The iterator gives you a `SRT`, but `SRT` has no public member or method, so it's not usable from outside the library.

tomaka commented:
Also I think that there should be an alternative API instead of just the iterator. The iterator is stateful, which is a big drawback in some situations.

I think you should provide a function on the skeleton to query any timer of any animation, and then just implement the animation iterator as a wrapper around this.

tafia commented:
I've added `SkinAnimation` struct to handle interpolations at any time.
I think it should not be for the skeleton to support it because of the skin/animation name searches.

``` rust
let slots = Skeleton::from_reader(reader)
                .get_animated_skin("default", Some("walk")).interpolate(0.55);
// or
for slots in Skeleton::from_reader(reader)
                .get_animated_skin("default", Some("walk")).iter(0.01) {
    // ...
}
```

tafia commented:
By the way, to test it, I've tried creating a glium project based on your tutorial.
https://github.com/tafia/spine-render

Unfortunately, I couldn't have transparent png image... 
If you have 2 mins can you help me and tell me what I'm doing wrong ?

tomaka commented:
When drawing, instead of passing `Default::default()` you need to pass something like this:

```
glium::DrawParameters {
    blend: glium::Blend::alpha_blending(),
    .. Default::default()
}
```

I'm currently converting my code to the new design to see if it works.

tafia commented:
Crazy it's that simple. You can't know how long I've searched for it ...
Thanks !!!

Are the last changes OK for you ?
As I haven't "see" it working I can't guarantee anything unfortunately, i'm
afraid you're making tests too soon ...
On 19 Oct 2015 20:42, "tomaka" notifications@github.com wrote:

> When drawing, instead of passing Default::default() you need to pass
> something like this:
> 
> glium::DrawParameters {
>     blend: glium::Blend::alpha_blending(),
>     .. Default::default()
> }
> 
> I'm currently converting my code to the new design to see if it works.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/tomaka/spine-rs/pull/6#issuecomment-149202660.

tomaka commented:
Yeah the changes look good.

There are two or three minor annoyances that I think could be fixed, but I'll wait until this PR is merged.

tomaka commented:
I got a linking error while trying your code: https://github.com/rust-lang/rust/issues/29164

tafia commented:
As stated in the rust issue, the error came because of some (a lot) of `struct`s not being publicly exposed.

I've decided to put relevant modules as public instead of reexporting only the `struct` I want.
The reason is that I could use the compiler to tell me what was wrong. I add to put a lot of comments. 

Most comments need to be rewritten but I'll do it later, it is in my todo list along with a proper documentation with examples. 

Some publicly available `fn`/`struct`s can probably remain private to simplify library usage.

I've also renamed `CalculatedSlot` into `Sprite`. I'm not very familiar with all these names (slot, spine, sprite, attachment ...) so I'm not really sure they fit properly.

tomaka commented:
I tried your code in my game (with the revision before your latest commits), and nothing shows up at all. I think it's because the positions are wrong.

At this point I think it's much easier to make it work with https://github.com/tafia/spine-render first, instead of me debugging it inside my complex code.

tafia commented:
Yes I agree.
I am very sorry to waste your time. The attachment size was never taken into account ...

I'm already making changes on my branch [betterSprite](https://github.com/tafia/spine-rs/tree/betterSprite).
In particular, I'm internalizing positions calculation in spine-rs (instead of returning srt, I return positions). I don't know if all these calculations should be on gpu side or not but I find it easier that way. I also decided not to use any math library because on 2D calculations are straightforward so it is simpler to debug.

Basic tests show good results (I mean, sprite positions are showing expected values on simple tests).
I couldn't try spine-render yet because I'm not at home and I got a compile error on my windows tablet (i686-pc-windows-gnu).

But don't waste your time, I'll let you know once I manage to get a walking spineboy! Hopefully it should be soon!

tafia commented:
Tried it ... Still not good but at least I got something!
![screenshot from 2015-10-21 17-20-57](https://cloud.githubusercontent.com/assets/7351502/10632634/42859828-7818-11e5-9601-b4340a4dbeea.png)

tafia commented:
Some more news as I think a new milestone has been reached.

I fixed I think the biggest bugs on `betterSprite` branch. It is still not perfect for several reasons but at least it can now:
- display a setup pose that looks like a sane spineboy (`animation = None`)
- kind of animate the skeleton
  - all bones look attached together during the animation so spineboy looks alive
  - i don't know exactly how it should render but the animations I've tested didn't look natural at all, so I suspect something is still wrong ...

I haven't benchmarked it recently but I believe it is still at least 4x better than current implementation. I already spoted several directions to improve it further.

Concerning spine-render, I don't know how to change sprite order on the fly ... so I ended up recreating a indiceBuffer everytime. I think it is really not efficient but couldn't spend time to find a better solution.

tafia commented:
Animations are now working properly I think. **You can probably try it on your project** if you don't change your attachment.

There are some more optimizations I'd like to try before intergating. Plus I haven't dealt with attachment name change yet.

tafia commented:
Attachment animation working too now!!

tomaka commented:
The fact that a sprite produces a list of positions instead of a SRT or a matrix is really annoying.
A game has to take into account the object's position, rotation, scale from the game logic as well, plus the position and rotation of the camera.

I had to manually apply the object's and camera's matrices to your positions in order to make it work, and I was lucky that my game manually uploads the sprites positions. Another engine could have uploaded matrices in a uniform buffer for example, in which case have positions is very unpractical.

Other than that, it works, which is great!

There's another problem though: I need to know how long an animation is. I'd like to call `interpolate(time % animation_len)` in order to repeat an animation forever, but I don't have access to that `animation_len`.

tafia commented:
> The fact that a sprite produces a list of positions instead of a SRT or a matrix is really annoying.
> A game has to take into account the object's position, rotation, scale from the game logic as well, plus the position and rotation of the camera.

Sorry to have changed that, was simpler for my tests/debug.
On my spine_render test, I manually modify the vertices positions. I thought the game logic could be modifying all these vertices at once (with a uniform) _after_ having setup them.

Truth is I lack a good example to see how unconvenient it is in real world (or how bad practice it is to modify buffers).

I'm not sure what is the best strategy about the initialization step (whether the attachment srt must be included into the calculated srt or taken once for all at initialization). I.e. Can `get_attachment_transformation` fn be used once for all at initialization instead of manually multiplying it everytime as done in master ?

> There's another problem though: I need to know how long an animation is. I'd like to call interpolate(time % animation_len) in order to repeat an animation forever, but I don't have access to that animation_len.

I think just putting the `SkinAnimation.duration` public should be enough. Else the `.iter().cycle()` iterator should work too (the iterator creation is _very_ cheap).

tomaka commented:
The biggest bottleneck in 2D is the upload speed. Uploading a uniform takes time. Uploading vertices takes time. If you upload both a uniform and vertices, you're paying twice the cost.

Of course that's negligible just for a few sprites, but when you design an engine you take this into account. Either you do all the transformations on the CPU and you upload only the vertices, or you do all the transformations on the GPU with a uniform (and the vertices are pre-uploaded).

Since you give out positions, you force anyone who uses the library to use the first solution.

tafia commented:
Here is a branch which is supposed to revert the `srt`s.
https://github.com/tafia/spine-rs/tree/skeleton-srt

I've added a method to get setup positions for all skin attachments:

``` rust
// get all attachments for the skin to preload vertices
skeleton.get_skin(....).map(|s| s.attachment_positions());
```

tafia commented:
I've tested it all on [spine-render](https://github.com/tafia/spine-render).
It works as expected so I merged the skeleton-srt branch.
Should be good now I suppose. Let me know!

tomaka commented:
Don't put `duration` as public. Add a getter instead. You don't want people to modify the duration.

tafia commented:
Done

I have 2 questions for you
1. I want to clean old code: Is there anything special you see missing in skeleton that was in master?
2. After implementing spine-render, there is still one area I think can be improved: attachment names. Currently both in the library and on the spine-render, I must search for the particular attachment _name_. It should be simpler instead to return a mere attachment _index_ representing the index of the attachment from the list `attachment_positions` returns (changing it to `Vec` instead of `HashMap` of course) 
   - this index will map easily to vertices and indices
   - on library side no more `String.clone()`

Apart that my todo list is still pretty huge. Performance-wise do you see any big issue?

tomaka commented:
What's missing from master is the function that returns the list of all possible attachments. This is very useful in order to know what should be loaded during a loading screen.

There's one thing that I think could be improved performance-wise: the `interpolate` function and the animation iterators should not return a `Vec`. It should be possible to create a custom iterator for the list of attachments, and this would avoid an allocation and a possible cache miss.

For the attachment names, this is a complicated topic. I suggest to merge this first and see later.

tafia commented:
> What's missing from master is the function that returns the list of all possible attachments. This is very useful in order to make sure that everything is loaded.

Ok I'll do it. Do you want the setup positions as well (as done with the `skin.attachment_positions`, which eventually could be more complicated than a rectangle) ?

> There's one thing that I think could be improved performance-wise: the interpolate function and the animation iterators should not return a Vec. It should be possible to create a custom iterator for the list of attachments, and this would avoid an allocation and a possible cache miss.

This should be easy to do. As a start all bone transformations will still be computed, and only the slots will be iterated.

tafia commented:
2 points addressed.
I am not 100% sure how it compares on benches because iterators may be optimized out.

tafia commented:
Fyi, I've started removing old code (and updating documentation) on the clean branch:
https://github.com/tafia/spine-rs/tree/clean

I think it is time to stress it now.

tafia commented:
I've merged several changes to the skeleton branch:
- clean: add all missing getters, updated examples and readme, rewrite atlas, add `to_matrix` functions, remove old code and unused dependencies
- performance (breaking changes): on my system went from 300 us to 180 us
  - have sprite and attachment timelines return `&str` attachment name instead of `String`,
  - change all color types to `[u8; 4]` instead of `Vec<u8>`
- rebased

Tell me if you want anything else. I'll squash all commits this week to have something more digest.

tomaka commented on Cargo.toml:
```diff
@@ -1,7 +1,8 @@
 name = "spine_tiny"
-version = "0.0.1"
-authors = [ "pierre.krieger1708@gmail.com" ]
+version = "0.0.2"
```
Just bump to 0.1.0

Versions like `0.0.x` are bad.

tafia commented:
done

tafia committed:
Skeleton improvements

- Have many structs initialized when loading json
- Better bezier calculations
- Precompute as many data as possible
  - SkinAnimation: search for all references when initializing
  - Sprites: Iterator over sprites
- Atlas files management

tafia commented:
All squashed!

tomaka commented:
Thanks!

tafia commented:
Thank YOU
That was fast!
On 15 Nov 2015 17:09, "tomaka" notifications@github.com wrote:

> Merged #6 https://github.com/tomaka/spine-rs/pull/6.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/tomaka/spine-rs/pull/6#event-464732509.
